src/map/script.cpp

```diff
@@ -6659,6 +6659,189 @@ BUILDIN_FUNC(getelementofarray)
 	return SCRIPT_CMD_SUCCESS;
 }
 
+/// Return the index number of the value in an array.
+/// ex: inarray arr,1;
+///
+/// inarray <array variable>,<value>;
+BUILDIN_FUNC(inarray)
+{
+	struct script_data *data;
+	const char* name;
+	int id, i, array_size;
+	struct map_session_data* sd = NULL;
+	struct reg_db *ref = NULL;
+	data = script_getdata(st, 2);
+
+	if (!data_isreference(data))
+	{
+		ShowError("buildin_inarray: not a variable\n");
+		script_reportdata(data);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	name = reference_getname(data);
+	ref = reference_getref(data);
+
+	if (not_server_variable(*name) && !script_rid2sd(sd))
+		return SCRIPT_CMD_FAILURE;
+
+	array_size = script_array_highest_key(st, sd, name, ref) - 1;
+
+	if (array_size > SCRIPT_MAX_ARRAYSIZE)
+	{
+		ShowError("buildin_inarray: The array is too large.\n");
+		script_reportdata(data);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	id = reference_getid(data);
+	if (is_string_variable(name))
+	{
+		const char* temp;
+		const char* value;
+		value = script_getstr(st, 3);
+		for (i = 0; i <= array_size; ++i)
+		{
+			temp = (char*)get_val2(st, reference_uid(id, i), ref);
+			script_removetop(st, -1, 0);
+			if (!strcmp(temp, value))
```

Comment:
temp never will be set.
Because at line above
```
script_removetop(st, -1, 0);
```

you're already removing data from stack.
You can try to test on any string array, and will see results.

Change to:
```cpp
	if (is_string_variable(name))
	{
		const char* temp;
		const char* value;
		value = script_getstr(st, 3);
		for (i = 0; i <= array_size; ++i)
		{
			temp = (char*)get_val2(st, reference_uid(id, i), ref);
			if (!strcmp(temp, value))
			{
				script_removetop(st, -1, 0);
				script_pushint(st, i);
				return SCRIPT_CMD_SUCCESS;
			}
			script_removetop(st, -1, 0);
			
		}
	}
```

the same do for interger, and for countinarray, you can test, if do not trust :D

Reply:
isn't doing this would remove only the last one from the stack ?

so i would need to change the -1 to what ever the arraysize when the for loop break 

i am not sure but i thing this code may send all the values in the array to the script
if you try to `mes` it it will write all the array elements -1
correct me if i am wrong please

Reply:
@anacondaqq to clear things up:
By using get_val2 the variable name is pushed onto the top of the stack before the value is resolved.
The actual value behind the variable stays untouched.
By removing the top stack entry after the call all you do is removing the variable name again. So it does not matter when you do it - you just have to do it.